allauth/socialaccount/providers/apple/views.py

```diff
@@ -0,0 +1,180 @@
+
+        public_key = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(apple_public_key))
+        return public_key
+
+    def get_client_id(self, provider):
+        app = SocialApp.objects.get(provider=provider.id)
+        return [aud.strip() for aud in app.client_id.split(",")]
+
+    def get_verified_identity_data(self, id_token):
+        provider = self.get_provider()
+        allowed_auds = self.get_client_id(provider)
+
+        try:
+            public_key = self.get_public_key(id_token)
+            identity_data = jwt.decode(
+                id_token,
+                public_key,
+                algorithms=["RS256"],
+                verify=True,
+                audience=allowed_auds,
+                issuer="https://appleid.apple.com",
+            )
+            return identity_data
+
+        except jwt.PyJWTError as e:
+            raise OAuth2Error("Invalid id_token") from e
+
+    def parse_token(self, data):
+        token = SocialToken(
+            token=data["access_token"],
+        )
+        token.token_secret=data.get("refresh_token", "")
+
+        expires_in = data.get(self.expires_in_key)
+        if expires_in:
+            token.expires_at = (
+                timezone.now() + timedelta(seconds=int(expires_in))
+            )
+
+        # `user_data` is a big flat dictionary with the parsed JWT claims
+        # access_tokens, and user info from the apple post.
+        identity_data = self.get_verified_identity_data(data["id_token"])
+        token.user_data = {**data, **identity_data}
+
+        return token
+
+    def complete_login(self, request, app, token, **kwargs):
+        extra_data = token.user_data
+        login = self.get_provider().sociallogin_from_response(
+            request=request, response=extra_data
+        )
+        login.state["id_token"] = token.user_data
+
+        # We can safely remove the apple login session now
+        # Note: The cookie will remain, but it's set to delete on browser close
+        try:
+            request.apple_login_session.delete()
+        except AttributeError:
+            pass
+
+        return login
+
+    def get_user_scope_data(self, request):
+        user_scope_data = request.apple_login_session.get("user", "")
+        try:
+            return json.loads(user_scope_data)
+        except json.JSONDecodeError:
+            # We do not care much about user scope data as it maybe blank
+            # so return blank dictionary instead
+            return {}
+
+    def get_access_token_data(self, request, app, client):
+        """ We need to gather the info from the apple specific login """
+        add_apple_session(request)
+
+        # Exchange `code`
+        code = get_request_param(request, "code")
+        access_token_data = client.get_access_token(code)
+
+        return {
+            **access_token_data,
+            **self.get_user_scope_data(request),
+            "id_token":request.apple_login_session.get("id_token")
```

Comment:
The contents of `access_token_data` should be the response from `POST https://appleid.apple.com/auth/token` (see `AppleOAuth2Client.get_access_token()`). According to [Apple's documentation](https://developer.apple.com/documentation/sign_in_with_apple/tokenresponse), the response should already contain the `id_token` attribute, and that is in fact what I see. So why override the `id_token` key with what is stored in `request.apple_login_session`?

This caused a problem from me when using ngrok to test apple login on localhost. Even though ngrok provides a public https URL, I was being directed back to the non-https version during the Apple login flow for reasons that I don't yet understand. Because of this, `request.apple_login_session` was being lost because `persist_apple_session()` stores the cookie with `secure=True`.

I can work around this issue on my local in two ways:
(1) comment out the above line of code.
(2) comment out `secure=True` in the `persist_apple_session()` function.

It may be worth considering whether the above line of code is necessary, given that `access_token_data` should already contain a value for `id_token`. If there _are_ scenarios in which `access_token_data` might be missing a value for `id_token`, this portion of code could be adjusted to first check if `id_token` is present in `access_token_data` before overriding it.

Reply:
I too have found trouble with this behaviour. Don't understand why we override `id_token`

Reply:
If I had seen your reviews a few days ago...
"Social Network Login Failure" says almost nothing to me.
Tracing up the reason why allauth missed `id_token` despite of valid `id_token` in my chrome network tap, was so hard journey.

2021.01.17
`Django==2.2.9`
`PyJWT==1.7.1`
At `HTTP` production environment,
this review really works for me and I finally made apple sign in successfully.

Thank you!  ☺️